src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -791,3 +930,341 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+ * reach a new node altogether). In the context of single-statement commands,
+ * GetConnectionForPlacement simply falls through to  GetOrEstablishConnection.
+ */
+static PGconn *
+GetConnectionForPlacement(ShardPlacement *placement, bool isModificationQuery)
+{
+	NodeConnectionKey participantKey;
+	NodeConnectionEntry *participantEntry = NULL;
+	bool entryFound = false;
+
+	/* if not in a transaction, fall through to connection cache */
+	if (xactParticipantHash == NULL)
+	{
+		PGconn *connection = GetOrEstablishConnection(placement->nodeName,
+													  placement->nodePort);
+
+		return connection;
+	}
+
+	Assert(IsTransactionBlock());
+
+	MemSet(&participantKey, 0, sizeof(participantKey));
+	strlcpy(participantKey.nodeName, placement->nodeName, MAX_NODE_LENGTH + 1);
+	participantKey.nodePort = placement->nodePort;
+
+	participantEntry = hash_search(xactParticipantHash, &participantKey, HASH_FIND,
+								   &entryFound);
+
+	if (entryFound)
+	{
+		if (isModificationQuery)
+		{
+			RecordShardIdParticipant(placement->shardId, participantEntry);
+		}
+
+		return participantEntry->connection;
+	}
+	else
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_DOES_NOT_EXIST),
+						errmsg("no transaction participant matches %s:%d",
+							   placement->nodeName, placement->nodePort),
+						errdetail("Transactions which modify distributed tables may only "
+								  "target nodes affected by the modification command "
+								  "which began the transaction.")));
+	}
+}
+
+
+/*
+ * PurgeConnectionForPlacement provides a way to purge an invalid connection
+ * from all relevant connection hashes using the placement involved in the
+ * query at the time of the error. If a transaction is ongoing, this function
+ * ensures the right node's connection is set to NULL in the participant map
+ * for the transaction in addition to purging the connection cache's entry.
+ */
+static void
+PurgeConnectionForPlacement(ShardPlacement *placement)
+{
+	NodeConnectionKey nodeKey;
+	char *currentUser = CurrentUserName();
+
+	MemSet(&nodeKey, 0, sizeof(NodeConnectionKey));
+	strlcpy(nodeKey.nodeName, placement->nodeName, MAX_NODE_LENGTH + 1);
+	nodeKey.nodePort = placement->nodePort;
+	strlcpy(nodeKey.nodeUser, currentUser, NAMEDATALEN);
+
+	PurgeConnectionByKey(&nodeKey);
+
+	if (xactParticipantHash != NULL)
+	{
+		NodeConnectionEntry *participantEntry = NULL;
+		bool entryFound = false;
+
+		Assert(IsTransactionBlock());
+
+		MemSet(&nodeKey.nodeUser, 0, sizeof(nodeKey.nodeUser));
+		participantEntry = hash_search(xactParticipantHash, &nodeKey, HASH_FIND,
+									   &entryFound);
+
+		Assert(entryFound);
+
+		participantEntry->connection = NULL;
+	}
+}
+
+
+static void
+RecordShardIdParticipant(uint64 affectedShardId, NodeConnectionEntry *participantEntry)
+{
+	XactShardConnSet *shardConnSetMatch = NULL;
+	ListCell *listCell = NULL;
+	MemoryContext oldContext = NULL;
+	List *connectionEntryList = NIL;
+
+	foreach(listCell, xactShardConnSetList)
```

Comment:
Agreed. I initially created another hash, but it was quite onerous and given that this is _O(xact_shards * replication)_ it seems likely this will often run just as fast as any hash or tree.